docs/madr/decisions/031-automatic-rechable-services.md

```diff
@@ -0,0 +1,289 @@
+    * but it also can be undesirable (e.g. github will 403 on repositories you don't have access to if you **are** part of the organisation)
+
+## Pros and Cons of the Options
+
+### Using MeshTrafficPermission to generate reverse graph that would feed into reachable services
+
+#### Mapping
+
+MeshTrafficPermission is defined by the service owner and states which services can communicate with that service.
+Based on that information we could build a reverse graph containing information who can talk to whom and based on that we could trim the configs created (which would be equivalent to populating reachable services field).
+
+So MTPs like this:
+
+```yaml
+apiVersion: kuma.io/v1alpha1
+kind: MeshTrafficPermission
+metadata:
+  namespace: kuma-system
+  name: mtp-b
+spec:
+  targetRef:
+    kind: MeshService
+    name: d
+  from:
+    - targetRef:
+        kind: MeshService
+        name: a
+      default:
+        action: Allow
+    - targetRef:
+        kind: MeshService
+        name: b
+      default:
+        action: Allow
+    - targetRef:
+        kind: MeshService
+        name: c
+      default:
+        action: Allow
+```
+
+Would translate into:
+
+```mermaid
+graph TD
+    A-->D
+    B-->D
+    C-->D
+```
+
+Let's consider other `targetRef` with allow types in combination with `top`/`from` level:
+1. `Mesh/Mesh` - this won't help us because that's the default, anyone can talk to anyone and we can't
+
+2. `MeshService/Mesh`  
+
+```yaml
+apiVersion: kuma.io/v1alpha1
+kind: MeshTrafficPermission
+metadata:
+  namespace: kuma-system
+  name: mtp-ms-to-mesh
+spec:
+  targetRef:
+    kind: MeshService
+    name: a
+  from:
+    - targetRef:
+        kind: Mesh
+      default:
+        action: Allow
+```
+
+```mermaid
+graph TD
+    B-->A
+    C-->A
+    D-->A
+```
+
+We could either define an edge between every service to that service or have a special list that contains allIncomingAllowed.
+Services from that list would always be added as an outgoing edge.
+
+- `Mesh/MeshService`
+
+```yaml
+apiVersion: kuma.io/v1alpha1
+kind: MeshTrafficPermission
+metadata:
+  namespace: kuma-system
+  name: mtp-mesh-to-ms
+spec:
+  targetRef:
+    kind: Mesh
+  from:
+    - targetRef:
+        kind: MeshService
+        name: a
+      default:
+        action: Allow
+```
+
+```mermaid
+graph TD
+    A-->B
+    A-->C
+    A-->D
+```
+
+This is simply `A` can talk to anyone, this could be either adding and outgoing edge from this service to every other service or having it on a list called allOutgoing and not bothering to trim the configuration.
+
+3. `MeshService/MeshService` - already talked about, this is the easy case
+
+4. Subsets 
+
+There is a strong case that makes sense using `k8s.kuma.io/namespace: xyz` to allow only traffic from a specific namespace.
+
+We will support `MeshSubset` and `MeshServiceSubset` on the `from` level and support `k8s.kuma.io/*` labels on the `top` level.
```

Comment:
@jakubdyszkiewicz maybe you can fill in the details on this if needed.

Reply:
Not sure if there is much to add here. Maybe that we could support all common labels (not only k8s.kuma.io/) between all instances of the Service, but at the higher computation cost, so we want to avoid this for now.